src/sentry/issues/issue_velocity.py

```diff
@@ -0,0 +1,189 @@
+                        "divide",
+                        [
+                            Column("past_week_event_count"),
+                            Function(
+                                "dateDiff", ["hour", Column("first_seen"), now]
+                            ),  # otherwise divide by its age in hours
+                        ],
+                    ),
+                ],
+                "hourly_event_rate",
+            ),
+        ],
+        groupby=[Column("group_id")],
+        where=[
+            Condition(
+                Column("timestamp"), Op.GTE, ninety_days_ago
+            ),  # include issues up to the oldest retention date to determine whether an issue is older than the week or not
+            Condition(Column("timestamp"), Op.LT, now),
+            Condition(Column("project_id"), Op.EQ, project.id),
+        ],
+        having=[
+            Condition(Column("past_week_event_count"), Op.GT, 1)
+        ],  # exclude any issues that had only 1 event in the past week
+        orderby=[OrderBy(Column("first_seen"), Direction.ASC)],
+        limit=Limit(10000),
+    )
+
+    query = Query(
+        match=subquery,
+        select=[
+            Function(
+                THRESHOLD_QUANTILE["function"],
+                [Column("hourly_event_rate")],
+                THRESHOLD_QUANTILE["name"],
+            )
+        ],  # get the approximate 90th percentile of the event frequency in the past week
+        limit=Limit(1),
+    )
+
+    request = Request(
+        dataset=Dataset.IssuePlatform.value,
+        app_id=REFERRER,
+        query=query,
+        tenant_ids={"referrer": REFERRER, "organization_id": project.organization.id},
+    )
+
+    result = raw_snql_query(request, referrer=REFERRER)["data"]
+    if len(result) == 0:
+        return None
+
+    try:
+        return result[0][THRESHOLD_QUANTILE["name"]]
+    except KeyError:
+        logger.exception(
+            "Unexpected shape for threshold query results",
+            extra={"project_id": project.id, "results_received": json.dumps(result)},
+        )
+        return None
+
+
+def update_threshold(project: Project, threshold_key: str, stale_date_key: str) -> Optional[float]:
+    """
+    Runs the calculation for the threshold and saves it and the date it is last updated to Redis.
+    If no threshold could be calculated for whatever reason, we don't save anything to Redis.
+    """
+    threshold = calculate_threshold(project)
+    if threshold is None:
+        logger.error(
+            "Velocity threshold couldn't be calculated, query returned nothing",
+            extra={"project_id": project.id},
+        )
+        return None
+    elif math.isnan(threshold):  # indicates there were no valid events to base the calculation
+        return None
+
+    client = get_redis_client()
+    with client.pipeline() as p:
+        p.set(threshold_key, threshold, ex=REDIS_TTL)
+        p.set(stale_date_key, datetime.utcnow().timestamp(), ex=REDIS_TTL),
```

Comment:
alternative is to do `p.mset` to set both keys at once, but then two calls of `p.expire` (couldn't find a way to bulk set expiration)